File: system_proxy/android/src/main/kotlin/com/nove8/system_proxy_service/SystemProxyServicePlugin.kt
package com.nove8.system_proxy_service

import com.nove8.plugins.common.CommonPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Flutter plugin that reports the HTTP proxy configured in the system settings.
 *
 * Android exposes the Wi-Fi proxy of the active network through the standard
 * JVM proxy system properties.
 *
 * Extends [CommonPlugin] to use standardized channel naming and lifecycle management.
 */
class SystemProxyServicePlugin : CommonPlugin(CHANNEL_NAME) {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            GET_SYSTEM_PROXY_METHOD -> result.success(getSystemProxy())
            else -> super.onMethodCall(call, result)
        }
    }

    private fun getSystemProxy(): Map<String, Any?> {
        return mapOf(
            HOST_PARAM to System.getProperty(PROXY_HOST_PROPERTY),
            PORT_PARAM to System.getProperty(PROXY_PORT_PROPERTY)?.toIntOrNull(),
        )
    }

    private companion object {
        // Channels
        private const val CHANNEL_NAME = "system_proxy"

        // Methods
        private const val GET_SYSTEM_PROXY_METHOD = "getSystemProxy"

        // Params
        private const val HOST_PARAM = "host"
        private const val PORT_PARAM = "port"

        // Properties
        private const val PROXY_HOST_PROPERTY = "http.proxyHost"
        private const val PROXY_PORT_PROPERTY = "http.proxyPort"
    }
}

File: system_proxy/ios/system_proxy_service/Sources/system_proxy_service/SystemProxyServicePlugin.swift
import Flutter
import common_plugin

/**
 Flutter plugin that reports the HTTP proxy configured in the system settings.

 iOS exposes the proxy of the active network through `CFNetworkCopySystemProxySettings`.
 Only a manually configured proxy is reported, an automatic configuration (PAC) is ignored.
 */
public class SystemProxyServicePlugin: CommonPlugin, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        CommonPlugin.register(with: registrar, plugin: SystemProxyServicePlugin())
    }

    public override class var pluginKey: String { "SystemProxyServicePlugin" }

    public override var channelName: String { SystemProxyServicePlugin.CHANNEL_NAME }

    public override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case SystemProxyServicePlugin.GET_SYSTEM_PROXY_METHOD:
            result(getSystemProxy())
        default:
            super.handle(call, result: result)
        }
    }

    private func getSystemProxy() -> [String: Any?] {
        guard let settings = CFNetworkCopySystemProxySettings()?.takeRetainedValue() as? [String: AnyObject],
              settings[SystemProxyServicePlugin.HTTP_ENABLE_SETTING] as? Bool == true
        else {
            return [:]
        }

        return [
            SystemProxyServicePlugin.HOST_PARAM: settings[SystemProxyServicePlugin.HTTP_PROXY_SETTING] as? String,
            SystemProxyServicePlugin.PORT_PARAM: settings[SystemProxyServicePlugin.HTTP_PORT_SETTING] as? Int,
        ]
    }

    // Channels
    private static let CHANNEL_NAME = "system_proxy"

    // Methods
    private static let GET_SYSTEM_PROXY_METHOD = "getSystemProxy"

    // Params
    private static let HOST_PARAM = "host"
    private static let PORT_PARAM = "port"

    // Settings
    private static let HTTP_ENABLE_SETTING = "HTTPEnable"
    private static let HTTP_PROXY_SETTING = "HTTPProxy"
    private static let HTTP_PORT_SETTING = "HTTPPort"
}
